Add `NumberControl` to support min/max font size for `Tag Cloud Block`

amustaque97 opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->
Fixes: #37166
<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
Add `NumberControl` input to enter minimum and maximum font size value. It will help user to decide how they want tag cloud to look. Earlier it used to modify fontsize based on the number of post for a particular tag.

**NOTE: When I reload the page these values reset to default values.** 😞 

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

1. Checkout to this branch
2. Create multiple tags here `http://localhost:8888/wp-admin/edit-tags.php?taxonomy=post_tag`
3. Create multiple different post and add tags created in the above step.
4. Create/edit a post and add `/tag cloud` block.
5. For `/tag cloud` block go to settings, there you can see 2 new input has been added to control font size.
6. Adjust/modify those numbers and save the post.
7. Open same post that has `/tag cloud` block and check its font size.

## Screenshots <!-- if applicable -->
- NumberControl
![image](https://user-images.githubusercontent.com/21127788/145701748-29aef42d-4eda-4b8d-980a-b7e74eeb965e.png)

- After post publish
<img width="1155" alt="Screenshot 2021-12-12 at 11 00 10 AM" src="https://user-images.githubusercontent.com/21127788/145701598-741c1547-9216-4abc-9678-cc4993e7a415.png">


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Feature
## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

amustaque97 committed:
Add `NumberControl` to support min/max font size for `Tag Cloud Block`

amustaque97 committed:
Regenerate fixtures for `tag-cloud` block

ntsekouras commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -82,6 +89,20 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
+					}
+				/>
+				<NumberControl
+					label={ __( 'Largeest Font Size (Pt)' ) }
```
I guess we can make labels smaller like `Largest size(pt)`. Also note the typo in `Largest` here.

ntsekouras commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -82,6 +89,20 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
+					value={ largestFontSize }
+					onChange={ ( value ) =>
+						setAttributes( { largestFontSize: value } )
+					}
```
We should probably add some defaults, at least for start. Maybe something like the below? (🤔 )
```
min={ 8 }
max={ 60 }
```

amustaque97 replied:
Sure, agree with you.

ntsekouras commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -82,6 +89,20 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
 					max={ MAX_TAGS }
 					required
 				/>
+				<NumberControl
```
I think it will be better to have these controls in one row. You can wrap them with `<Flex>` component and then each `NumberControl` should be wrapped with `<FlexItem isBlock>` component.

amustaque97 replied:
Wrapping fields inside Flex looks more clean and consistent UI. Thanks for this. :smile:

ntsekouras reviewed (commented):
Thanks for your work @amustaque97! 💯 

This works well code wise! I left a couple of notes about making the controls a bit more compact and it would be ready to land. After the changes we would just need a design approval of course maybe by @jasmussen?

amustaque97 committed:
address review comment

In this commit, I have addressed following  comments:
1. fix label of the NumberControl field
2. add min and max value of NumberControl field
3. Wrap NumberControl fields inside flex to make UI consistent

ntsekouras commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -82,6 +91,30 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
+					</FlexItem>
+					<FlexItem isBlock>
+						<NumberControl
+							label={ __( 'Largest size (pt)' ) }
```
```suggestion
							label={ __( 'Largest size(pt)' ) }
```

ntsekouras commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -82,6 +91,30 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
+				<Flex>
+					<FlexItem isBlock>
+						<NumberControl
+							label={ __( 'Smallest size (pt)' ) }
```
```suggestion
							label={ __( 'Smallest size(pt)' ) }
```

amustaque97 committed:
Remove space in label

Co-authored-by: Nik Tsekouras <ntsekouras@outlook.com>

amustaque97 committed:
Remove space from label of smallest size input field

Co-authored-by: Nik Tsekouras <ntsekouras@outlook.com>

ntsekouras commented:
You need to add the changes to block docs since you changed block attributes. Run `npm run api-docs:ref` and `npm run api-docs:blocks`.

amustaque97 committed:
add smallestFontSize and largestFontSize in core blocks docs

amustaque97 committed:
fix NumberControl typecasting issue

This commit is done inorder to fix type issue since NumberControl returns
string and because of that It was not able to render font size properly.

jasmussen commented:
Thank you for the PR! Here's how the tag cloud looks in trunk:
<img width="1392" alt="before" src="https://user-images.githubusercontent.com/1204802/151513407-4534f2f2-38ab-40f8-af89-198f7f36ecb8.png">

Here's this PR applied:

<img width="1392" alt="after" src="https://user-images.githubusercontent.com/1204802/151513440-636b3ebf-04a1-4ee1-b230-bfc53bc811b8.png">

I thought about whether there was a better way to combine min/max font sizes into a singular control, but for comparison the paragraph font size looks like this:
<img width="364" alt="Typography panel states" src="https://user-images.githubusercontent.com/1204802/151513755-91e6d63b-4844-455e-8ced-805fd7864b3e.png">

So it's probably best to not diverge too far from this.

The following is a mockup that uses new components from the effort outlined in #27331, so it looks a bit different from what's shipping but it's what it could eventually look like:

<img width="364" alt="mockup" src="https://user-images.githubusercontent.com/1204802/151513879-d70ae60a-a366-42cb-a568-48122c951fa6.png">

But one single recommedation I'd make is to adopt the unit-selector input field. That would both let us remove the `(pt)` from the label, and pick other units. Is that feasible?

In smaller changes, if it isn't too much, I'd change the panel title from "Tag Cloud settings" to just "Settings" — there really isn't need for the redundancy. 

Nice work!

amustaque97 committed:
change font size units from `pt` to `px`

amustaque97 committed:
fix php lint error replace tabs with spaces during mid-line alignment

amustaque97 commented:
Thank you for the reply! I can change the units from `pt` to `px` but not sure how to add `px` keyword to the right side of the `NumberControl` input field. Maybe reviewers can suggest me something!

amustaque97 committed:
replace `NumberControl` with `InputControl`

this is done to be in sync with the UI illustrations and UI design pattern

amustaque97 commented:
I tried my best to implement UI as shown in the illustration. I testing it making mobile/ipad responsive. Here is the screenshot!
<img width="1095" alt="Screenshot 2022-01-28 at 3 58 48 PM" src="https://user-images.githubusercontent.com/21127788/151531810-e0ed7e2b-981d-4198-abe2-2e79acb1414c.png">

In order to implement this I have used `InputControl` instead of `NumberControl`

ntsekouras commented:
Thanks for the iterations here @amustaque97!

In order to implement properly what @jasmussen suggests you should use `UnitControl` component. What this means in practice is that:
1. You should update the new attributes to being `strings` with default values in `pt`
2. This will also require extra handling (splitting) these values in (value/unit) in order to be passed properly in `wp_tag_cloud` (php part).

We can't change the default behavior to `px` because it will change the existing blocks appearance.

This is a bit more extra work but provides more flexibility.

amustaque97 committed:
add new attribute `fontSizeUnit`

In this commit, a new attribute is introduced to select different font unit
example `px` or `pt`. Core blocks docs is also updated

amustaque97 committed:
replace `InputControl` with `UnitControl`

In this commit, `UnitControl` component is used to configure smallestFontSize
and largestFontSize. UnitControl offers input field with a select html tag.
`fontSizeUnit` single attribute is used to keep unit in both the fields in sync

amustaque97 committed:
update `wp_tag_cloud` args to handle `fontSizeUnit`

amustaque97 committed:
fix unit test

test cases were failing because of new attribute `fontSizeUnit`. It is fixed now.

ntsekouras commented on packages/block-library/src/tag-cloud/block.json:
```diff
@@ -28,6 +28,10 @@
 		"largestFontSize": {
 			"type": "number",
 			"default": 22
+		},
```
The sizes should be `strings` -  attribute (eg `largestFontSize`) with values like `8pt, 17px` etc... Check out other examples in the repo using the `UnitControl` component like `Column` and others.

ntsekouras commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -33,7 +36,19 @@ const MIN_TAGS = 1;
+		fontSizeUnit,
+	} = attributes;
+
+	const units = [
```
You should use `useCustomUnits` hook here.

ntsekouras replied:
This comment is still valid 😄

ntsekouras commented:
What occurred to me right now is that with unit control we can set different units per size (smallest, largest), but there is no support for this in `wp_tag_cloud`. So we can't go this way as is a case I don't think we've encountered before(design wise).

Maybe a first iteration of this could be to support only the `pt` as it was before the updates in the PR with the `px` related changes and find a design about this? 🤔 What do you think @jasmussen ?

amustaque97 commented:
@ntsekouras  Again! this was too quick 😅  💨 . TBH I was not ready for the review yet.  I was trying few things. Best thing to do right now is to wait for the reply from design team.

amustaque97 commented:
I would suggest is to make both `smallestFontSize` and `largestFontSize` unit in sync. Example if we are updating unit for any one of it. It should set the same unit for the other one. I was trying to implement the same using `useState` hooks. Umm. for some reason it is not able to sync correctly! Not sure whether it will be useful or not!

Adding a [link](https://codesandbox.io/s/weathered-frog-437u6?file=/src/App.js) for sample example though it is not an actual implementation but the idea would be same.

jasmussen commented:
Took this one for a spin:
![tag cloud](https://user-images.githubusercontent.com/1204802/152110053-951e08c0-f9f1-472c-9c36-06c44c325c26.gif)

This looks great to me! Thanks for the extra effort. I don't think we need to sync the two unit types, it adds only limited value and I could imagine in the future someone wanting a minimum font size in pixels, and a max font size in responsive `vw` units, so having the two unchained feels like a feature. 

Honestly this one is good to go pending just a code sanity check. Is there a reason we limit support to `px` and `pt` or could we support the full range of font size units? This wouldn't be a blocker for the PR, just a question.

amustaque97 commented:
`wp_tag_cloud` supports only single font size unit at any time. So if font size is not in sync. User will get confuse which font unit is currently in use. User may assume that there is a bug. This is another reason why initially I hardcoded `pt`.

jasmussen commented:
> wp_tag_cloud supports only single font size unit at any time. So if font size is not in sync. User will get confuse which font unit is currently in use. User may assume that there is a bug. This is another reason why initially I hardcoded pt.

Oh that's helpful context. 🤔 

The unit selector you've implemented, though, is useful, and it would be nice to keep it. How hard would it be to keep those two in sync after all?

ntsekouras commented:
>I would suggest is to make both smallestFontSize and largestFontSize unit in sync. Example if we are updating unit for any one of it. It should set the same unit for the other one. 

In order to do that I guess we would just handle this on `onChange` function of each input and conditionally update the other one (check the unit of both attributes). 

Noting that changing the two attributes(smallest, largest) into type `string` is still needed and we will not need a third attribute of the `unit`.

amustaque97 committed:
change `smallestFontSize` and `largestFontSize` attribute to string

amustaque97 committed:
add condition to update font unit

amustaque97 committed:
extract font unit from integer fontSize attribute

amustaque97 committed:
remove `fontSizeUnit` attribute

amustaque97 committed:
fix test cases after removing `fontSizeUnit` and str (smallest/largest)FontSize

amustaque97 committed:
fix php lint errors

amustaque97 commented:
Addressed review comments. Code wise it is functional but on UI it doesn't update when we update font unit. I wanted to use something like [unit](https://developer.wordpress.org/block-editor/reference-guides/components/unit-control/#unit-string) so that it will be easy to update UI if we are changing units at code level. 

Looking forward to your inputs Nik!

jasmussen commented:
Took it for a quick spin, and I'm seeing this:
![units](https://user-images.githubusercontent.com/1204802/152946690-93268e45-3a69-4dd2-a2a2-6f6fd0c54e7c.gif)

This looks great. Thank you for going the extra mile. This feels good to me.

I'd appreciate if Nik could give this one a quick code review, but the user experience is good! 👏

ntsekouras commented on packages/block-library/src/tag-cloud/index.php:
```diff
@@ -15,9 +15,12 @@
 function render_block_core_tag_cloud( $attributes ) {
 	$args      = array(
 		'echo'       => false,
+		'unit'       => substr( $attributes['smallestFontSize'], -2 ),
```
In general all this handling should be done by parsing the value to a `decimal` number and extract the unit to use as well. By going this way we can actually support all of units (also filtered by `theme.json` with `useCustomUnits` and there are units that are more than two characters.

amustaque97 replied:
Hi thank you for the comment, could you please elaborate a bit more. I didn't get it properly.

ntsekouras replied:
Since we are going this way to allow more `units`, we shouldn't restrict it to `pt and px` only. So for example if we have `rem`, the `slice( -2 )` would not work as expected. Check some usages of `useCustomUnits` where `useSetting` is also used. Example:
```
const units = useCustomUnits( {
	availableUnits: useSetting( 'spacing.units' ) || [ 'px', 'em', 'rem' ],
} );
```

amustaque97 replied:
We are using `useCustomUnits` hook!

ntsekouras commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -82,6 +96,54 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
+							label={ __( 'Smallest size' ) }
+							value={ smallestFontSize }
+							onChange={ ( value ) => {
+								const unit = value.slice( -2 );
```
Same goes here about the parsing with `-2`. We need to handle it more gracefully and dynamically.

amustaque97 replied:
Agree 💯  ! Addressed this comment.

ntsekouras commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -82,6 +96,54 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
+								value = value.slice( 0, -2 );
+								const newValue =
+									value && +value >= 8 ? +value : 8;
+								setAttributes( {
```
The passed new value is already a string with the unit attached, so there is no need to parse and reconstruct for the main attribute. In this case `smallestFontSize` for example.

amustaque97 replied:
I have introduced new function `onFontChange` common for both `smallestFontSize` and `largestFontSize`. I hope this time I did make mistake! 😅

ntsekouras reviewed (commented):
Thanks for iterating on this one once again @amustaque97! Great work!

I've left some comments, but it's getting really close!

amustaque97 committed:
implement `useCustomUnits` hooks to support different units

amustaque97 committed:
add `addFontChange` common function for `(smallest/largest)FontSize`

amustaque97 committed:
use `preg_replace` to handle font unit and pass integer font size `wp_tag_cloud`

ntsekouras commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -54,6 +77,22 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
 		return [ selectOption, ...taxonomyOptions ];
 	};
 
+	const onFontChange = ( fontSize, value ) => {
```
We should use the current API for this parsing `__experimentalParseUnit` from components package. Also have in mind that with the merging of [this one](https://github.com/WordPress/gutenberg/pull/38987/), the function will be renamed.

Also we can't check with `indexOf` as it will show wrong results with things like `em and rem`. An implementation could be something like this:
```
const onFontSizeChange = ( fontSize, value ) => {
	// eslint-disable-next-line @wordpress/no-unused-vars-before-return
	const [ quantity, newUnit ] = parseUnit( value );
	if ( ! Number.isFinite( quantity ) || quantity < 0 || quantity > 100 ) {
		return;
	}
	const updateObj = { [ fontSize ]: value };
	// We need to keep in sync the `unit` changes to both `smallestFontSize`
	// and `largestFontSize` attributes.
	Object.entries( {
		smallestFontSize,
		largestFontSize,
	} ).forEach( ( [ attribute, currentValue ] ) => {
		const [ _value, _unit ] = parseUnit( currentValue );
		// Only add an update if the other font size attribute has a different unit.
		if ( attribute !== fontSize && _unit !== newUnit ) {
			updateObj[ attribute ] = `${ _value }${ newUnit }`;
		}
	} );
	setAttributes( updateObj );
};
```

amustaque97 replied:
whoa this code is much much better than mine. Thanks for this... learnt `Object.entries` 😋

ciampo replied:
Wandering if instead of `quantity < 0 || quantity > 100` we could use the `min` and `max` props on `UnitControl` instead?

ntsekouras replied:
Unfortunately this doesn't work well in all cases. This is the related issue: https://github.com/WordPress/gutenberg/issues/33291. We could still set some min/max values here though, but I guess we should add as `min` something like `0.1` maybe?

ciampo replied:
Adding a `min` of `0.1` sounds sensible (just wondering if there's ever an instance where a user would like to set a value of `0`, but sounds like a hack that we shouldn't allow).

> This is the related issue: https://github.com/WordPress/gutenberg/issues/33291.

Thank you for the link! It's actually quite relevant in context of the recent TypeScript refactor that I'm attempting in https://github.com/WordPress/gutenberg/pull/38753

ntsekouras commented on packages/block-library/src/tag-cloud/index.php:
```diff
@@ -15,9 +15,12 @@
 function render_block_core_tag_cloud( $attributes ) {
 	$args      = array(
 		'echo'       => false,
+		'unit'       => preg_replace( '/\d+/', '', $attributes['smallestFontSize'] ),
```
All these might need escaping for sanitisation. Also check if in our code base some util exists or similar implementation for extracting the value/unit. It's a common use case and there might probably be somewhere.

amustaque97 replied:
Indeed sanitisation is required. I checked function in code reference but no luck. I asked in the `#core` channel and got [reply](https://wordpress.slack.com/archives/C02RQBWTW/p1645816392566379). I don't think there is any function available. Also, for sanitisation is it okay to use `sanitize_text_field` method?

ntsekouras replied:
--cc @getdave that did a lot of escaping recently 😄 

Also noting that we could have decimals, so the regex should take that into account.

getdave replied:
I don't think we need to escape the args to this core function. I'd be concerned if we were creating an arbitary HTML string but this one is generated by a Core function and the one thing I learnt from all the escaping I did, it that we should trust Core functions, especially those which have [available filters](https://github.com/WordPress/wordpress-develop/blob/8cc9e53847f8a46258d9e42c29f8d524ec0af592/src/wp-includes/category-template.php#L765).

amustaque97 replied:
I have updated the regex to handle decimal cases.

ntsekouras reviewed (commented):
Thanks for the iterations here @amustaque97!

The other thing that sometimes  doesn't update the `unit` in the control, is something that seems to be internal to `UnitControl`. The values are updated properly in the block and when the controls are rerendered everything seems fine. --cc @ciampo for this issue.

I can reproduce sometimes  when changing one font size and then change the unit of the other fontSize control.

https://user-images.githubusercontent.com/16275880/155703068-55db1829-3ec2-49e8-9618-a8afd59494ac.mov

amustaque97 committed:
refactor `onFontChange` function to `onFontSizeChange` function

ciampo commented:
> The other thing that sometimes doesn't update the unit in the control, is something that seems to be internal to UnitControl. The values are updated properly in the block and when the controls are rerendered everything seems fine. --cc @ciampo for this issue.

> I can reproduce sometimes when changing one font size and then change the unit of the other fontSize control.

Hey @ntsekouras! Out of curiosity, are you able to reproduce that bug also in Storybook, or is it just happening in the context of the Tag Clouds settings?

amustaque97 committed:
`tag-cloud` block php handle decimal values `unit` property

given a comment to handle decimal value case when passed to `unit`
property. I updated the regex to `/^\d*\.?\d*/`. This regex can parse
values like `123.3px` or `.33px` or `123px`.

ntsekouras commented:
>Hey @ntsekouras! Out of curiosity, are you able to reproduce that bug also in Storybook, or is it just happening in the context of the Tag Clouds settings?

If you add another `UnitControl` in the stories that update both values, you can reproduce as well. It has to do with  updating both values in one `onChange`..

ciampo commented:
> If you add another `UnitControl` in the stories that update both values, you can reproduce as well. It has to do with updating both values in one `onChange`..

Thank you for the explanation — I hadn't initially understood what the problem exactly was (syncing the unit change across `UnitControl`s), now it's much clearer.

After a quick look, here's what I think is causing the problem:

The code in this PR is keeping `UnitControl` instances in sync in order by changing unit programmatically, while by default the unit is changed by the user when interacting with the component in the browser. These 2 different behaviours are also referred to as _uncontrolled_ and _controlled_ components. (also cc'ing @sarayourfriend who also worked on the component to fact-check my hypothesis) 

Basically, if you want to use keep units in sync programmatically, you need to use `UnitControl` in a controlled way — the way to achieve this is to pass the `unit` prop explicitly to both instances of `UnitControl`

I went ahead and tried applying these changes

<details>
<summary>Click to expand</summary>

```diff
diff --git a/packages/block-library/src/tag-cloud/edit.js b/packages/block-library/src/tag-cloud/edit.js
index a8a7afaffc..a03ac31673 100644
--- a/packages/block-library/src/tag-cloud/edit.js
+++ b/packages/block-library/src/tag-cloud/edit.js
@@ -50,6 +50,13 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
 		largestFontSize,
 	} = attributes;
 
+	const [ smallestFontSizeQuantity, smallestFontSizeUnit ] = parseUnit(
+		smallestFontSize
+	);
+	const [ largestFontSizeQuantity, largestFontSizeUnit ] = parseUnit(
+		largestFontSize
+	);
+
 	const units = useCustomUnits( {
 		availableUnits: useSetting( 'spacing.units' ) || [
 			'%',
@@ -132,7 +139,8 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
 					<FlexItem isBlock>
 						<UnitControl
 							label={ __( 'Smallest size' ) }
-							value={ smallestFontSize }
+							value={ smallestFontSizeQuantity }
+							unit={ smallestFontSizeUnit }
 							onChange={ ( value ) => {
 								onFontSizeChange( 'smallestFontSize', value );
 							} }
@@ -142,7 +150,8 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
 					<FlexItem isBlock>
 						<UnitControl
 							label={ __( 'Largest size' ) }
-							value={ largestFontSize }
+							value={ largestFontSizeQuantity }
+							unit={ largestFontSizeUnit }
 							onChange={ ( value ) => {
 								onFontSizeChange( 'largestFontSize', value );
 							} }
```
</details>

and the controls seem to behave as expected

https://user-images.githubusercontent.com/1083581/156171956-4d062389-d291-49aa-b47e-82d070a7f785.mp4

amustaque97 commented:
@ciampo thank you for your inputs and taking time to have a look at the PR. My initial thought was on the same lines in the [comment](https://github.com/WordPress/gutenberg/pull/37311#issuecomment-1032234469) As per the docs `unit` attribute is deprecated. Not sure if we need to fix the docs or not.

ciampo commented:
> @ciampo thank you for your inputs and taking time to have a look at the PR. My initial thought was on the same lines in the [comment](https://github.com/WordPress/gutenberg/pull/37311#issuecomment-1032234469) As per the docs `unit` attribute is deprecated. Not sure if we need to fix the docs or not.

You make a good point re. the deprecation. We can use `unit` for now if it unblocks this PR, and I can have a deeper look in the meantime

ciampo commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -54,6 +78,28 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
+			}
+		} );
+		setAttributes( updateObj );
+	};
```
Small nit: renaming some of the variables for added clarity

```suggestion
	const onFontSizeChange = ( fontSizeLabel, newValue ) => {
		// eslint-disable-next-line @wordpress/no-unused-vars-before-return
		const [ quantity, newUnit ] = parseUnit( newValue );
		if ( ! Number.isFinite( quantity ) || quantity < 0 || quantity > 100 ) {
			return;
		}
		const updateObj = { [ fontSizeLabel ]: newValue };
		// We need to keep in sync the `unit` changes to both `smallestFontSize`
		// and `largestFontSize` attributes.
		Object.entries( {
			smallestFontSize,
			largestFontSize,
		} ).forEach( ( [ attribute, currentValue ] ) => {
			const [ currentQuantity, currentUnit ] = parseUnit( currentValue );
			// Only add an update if the other font size attribute has a different unit.
			if ( attribute !== fontSizeLabel && currentUnit !== newUnit ) {
				updateObj[ attribute ] = `${ currentQuantity }${ newUnit }`;
			}
		} );
		setAttributes( updateObj );
	};
```

ciampo commented:
> I can have a deeper look in the meantime

Opened #39148 with a potential fix

amustaque97 committed:
write descriptive variables names

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

amustaque97 commented:
> > I can have a deeper look in the meantime
> 
> Opened #39148 with a potential fix

Can we merge the above PR first, then I will take a merge from `trunk` branch. Woah! `UnitControl` will work in sync automatically. 😁😁 WDYT?

ciampo commented:
> In this case, since it looks like React Testing Library recommends using user-event most of the time in favour of fireEvent ([here in the docs](https://testing-library.com/docs/dom-testing-library/api-events/)), it sounds like it'll be worth it to add it in?

Sounds good to me, thank you for your patience

ciampo commented:
Update: #39148 was merged.

@amustaque97 , could you rebase (or merge `trunk` into) this PR, so @ntsekouras and I can double-check if the "unit sync" behavior has been fixed? Thank you!

amustaque97 commented:
> Update: #39148 was merged.
> 
> @amustaque97 , could you rebase (or merge `trunk` into) this PR, so @ntsekouras and I can double-check if the "unit sync" behavior has been fixed? Thank you!

Took merge from `trunk`. 
`UnitControl` is working as expected ✨✨  "Unit sync is working correct"

cc @ntsekouras @ciampo

ciampo commented:
I can confirm that the unit syncing between the two `UnitControls` is working as expected in the Tag Cloud Block.

Although at this point I would also recommend to wait until #38987 is merged (hopefully later today once all CI tests pass). This PR will need another rebase/merge to update (or rename) a couple of lines of code, but I think it's much easier this way than the other way around (#38987 is already quite large and is blocking other follow-up work until is merged)

amustaque97 commented:
Yeah I agree, If I remember correctly, `parseUnit` is renamed also. So after rebase I need to fix that as well.

ciampo commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -7,14 +7,23 @@ import { map, filter } from 'lodash';
 	RangeControl,
+	__experimentalUnitControl as UnitControl,
+	__experimentalUseCustomUnits as useCustomUnits,
+	__experimentalParseUnit as parseUnit,
```
```suggestion
	__experimentalParseQuantityAndUnitFromRawValue as parseQuantityAndUnitFromRawValue,
```

ciampo commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -54,6 +78,28 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
+			smallestFontSize,
+			largestFontSize,
+		} ).forEach( ( [ attribute, currentValue ] ) => {
+			const [ currentQuantity, currentUnit ] = parseUnit( currentValue );
```
```suggestion
			const [ currentQuantity, currentUnit ] = parseQuantityAndUnitFromRawValue( currentValue );
```

ciampo commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -54,6 +78,28 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
 
+	const onFontSizeChange = ( fontSizeLabel, newValue ) => {
+		// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+		const [ quantity, newUnit ] = parseUnit( newValue );
```
```suggestion
		const [ quantity, newUnit ] = parseQuantityAndUnitFromRawValue( newValue );
```

ciampo reviewed (commented):
@amustaque97 #38987 has been merged too, feel free to rebase and make the necessary changes (e.g. rename `parseUnit`)

amustaque97 committed:
rename `parseUnit` to `parseQuantityAndUnitFromRawValue`

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

amustaque97 committed:
rename `parseUnit` import name

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

amustaque97 committed:
rename `parseUnit` to `parseQuantityAndUnitFromRawValue`

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

ciampo commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -54,6 +78,28 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
 
+	const onFontSizeChange = ( fontSizeLabel, newValue ) => {
+		// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+		const [ quantity, newUnit ] = parseQuantityAndUnitFromRawValue( newValue );
```
```suggestion
		const [ quantity, newUnit ] = parseQuantityAndUnitFromRawValue(
			newValue
		);
```

ciampo commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -54,6 +78,28 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
+			smallestFontSize,
+			largestFontSize,
+		} ).forEach( ( [ attribute, currentValue ] ) => {
+			const [ currentQuantity, currentUnit ] = parseQuantityAndUnitFromRawValue( currentValue );
```
```suggestion
			const [
				currentQuantity,
				currentUnit,
			] = parseQuantityAndUnitFromRawValue( currentValue );
```

ciampo committed:
Apply suggestions from code review

Fix formatting

ciampo commented:
@ntsekouras @getdave I believe this PR is ready for a final review, could you have a look? Thank you!

getdave reviewed (commented):
Thanks for working on this. I haven't yet tested manually, but I'd love to see an e2e test added to cover both editor and front end.

One way to test the front end output is shown here

https://github.com/WordPress/gutenberg/blob/ab65bf431643e3cfc237b6634b68a01bae0b81f4/packages/e2e-tests/specs/editor/blocks/navigation.test.js#L883-L885

jorgefilipecosta commented on packages/block-library/src/tag-cloud/index.php:
```diff
@@ -15,9 +15,12 @@
 function render_block_core_tag_cloud( $attributes ) {
 	$args      = array(
 		'echo'       => false,
+		'unit'       => preg_replace( '/^\d*\.?\d*/', '', $attributes['smallestFontSize'] ),
```
Hi @amustaque97, thank you for your contribution the things work well from the functional point of view 👍 

I found a security vulnerability in this PR that allows a contributor role to output any CSS even disallowed rules that we need to fix before merging. So in the end a user may be able to use this to totally change the design of the website (e.g: with z-index, content plus SVG backgrounds even though its role does not allow that).

To reproduce for simple position case do the following steps:
- Create a new post and paste the following string on the code editor `<!-- wp:tag-cloud {"smallestFontSize":"8pt;position:fixed","largestFontSize":"22pt;position:fixed"} /-->`.
- Save the post and verify the position of the tag cloud is fixed while a contributor user should not be able to make elements contain fixed positions.

To fix this we can add a validation step with a regex stricter e.g: something like `/[0-9.]+[a-zA-Z%]+$` that forces a number followed by a set of alphabetic characters that are units but does not allow ";" or ".".  Or instead of preg_replace extract the units using preg_match and `[0-9.]+([a-zA-Z%]+)$`  where the units are part of group 1.

getdave replied:
@jorgefilipecosta How does this happen when both smallest and largest font size are passed through `floatVal`. 

```php
floatVal( $attributes['smallestFontSize'] )
```

This should obey [the rules of casting floats](https://www.php.net/manual/en/language.types.float.php#language.types.float.casting), namely:

> If the string is [numeric](https://www.php.net/manual/en/language.types.numeric-strings.php) or leading numeric then it will resolve to the corresponding float value, otherwise it is converted to zero (0).

What did I miss?

amustaque97 replied:
@jorgefilipecosta, I'm glad. thank you for taking a look at my PR and finding issues. I'm happy I learnt something new today. Would you mind taking a look at the fix again. 😅  cc @ciampo 

Thank you.

ntsekouras replied:
@getdave the issue is in [wp_generate_tag_cloud](https://developer.wordpress.org/reference/functions/wp_generate_tag_cloud/) in this line:
`esc_attr( str_replace( ',', '.', $tag_data['font_size'] ) . $args['unit'] )`

The `unit` that was passed, was the result of the regex handling, making to possible to add more css rules.

ntsekouras commented:
Thanks again for the iterations @amustaque97 and all the others with the help provided! 💯 

This PR LGTM is ready to go when the only remaining [security issue raised](https://github.com/WordPress/gutenberg/pull/37311#discussion_r821932973) by @jorgefilipecosta is addressed.

amustaque97 committed:
fix: security vulnerability add validate unit method for `wp_tag_cloud` args

amustaque97 committed:
fix php lint error using `composer format` command

ntsekouras commented on packages/block-library/src/tag-cloud/index.php:
```diff
@@ -13,11 +13,17 @@
  */
 function render_block_core_tag_cloud( $attributes ) {
+	$smallest_font_size = $attributes['smallestFontSize'];
+	$unit               = preg_match( '/[0-9.]+([a-zA-Z%]+)$/', $smallest_font_size ) ? preg_replace( '/^\d*\.?\d*/', '', $smallest_font_size ) : 'pt';
```
This is still unsafe 😄 

You can try this for example with `0.22em;position:fixed;top:0;left:0px`. We need a better regex here and maybe it would be safer to go the opposite way and declare the characters that we should not allow? 🤔  --cc @jorgefilipecosta

dd32 replied:
I would probably use `$unit = ( preg_match( '/^[0-9.]+(?P<unit>[a-z%]+)$/i', $smallest_font_size, $m ) ? $m['unit'] : 'pt' );` here.

It's explicit, requiring a `14px` or `0.22rem` value, but still defaulting to `pt` without too much hassle. It also avoids having to have multiple regular expressions.

amustaque97 replied:
I have made the required changes... :smile:

ciampo commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -54,6 +78,33 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
+		const [ quantity, newUnit ] = parseQuantityAndUnitFromRawValue(
+			newValue
+		);
+		if ( ! Number.isFinite( quantity ) || quantity < 0 || quantity > 100 ) {
```
With #39186 merged, would you be able to rebase on on top of latest `trunk` and see if the manual `quantity < 0 || quantity > 100` check can now swapped with `event.target?.validity.valid` (see [docs](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/number-control/README.md?plain=1#L100-L112))?

amustaque97 replied:
Oooh, I believe this is not required. In one of the review comment I was asked to use `UnitControl` instead of `NumberControl`. Thank you for sharing it 😄

ciampo replied:
`UnitControl` uses `NumberControl` under the hood — what I'm trying to say is that we should now able to use `min`/`max` instead of this manual check:

<details>
<summary>Click to expand</summary>


```diff
diff --git a/packages/block-library/src/tag-cloud/edit.js b/packages/block-library/src/tag-cloud/edit.js
index ae52d42f13..23515f1764 100644
--- a/packages/block-library/src/tag-cloud/edit.js
+++ b/packages/block-library/src/tag-cloud/edit.js
@@ -41,6 +41,9 @@ const MIN_TAGS = 1;
  */
 const MAX_TAGS = 100;
 
+const MIN_FONT_SIZE = 1;
+const MAX_FONT_SIZE = 100;
+
 function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
 	const {
 		taxonomy,
@@ -83,7 +86,7 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
 		const [ quantity, newUnit ] = parseQuantityAndUnitFromRawValue(
 			newValue
 		);
-		if ( ! Number.isFinite( quantity ) || quantity < 0 || quantity > 100 ) {
+		if ( ! Number.isFinite( quantity ) ) {
 			return;
 		}
 		const updateObj = { [ fontSizeLabel ]: newValue };
@@ -142,6 +145,8 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
 								onFontSizeChange( 'smallestFontSize', value );
 							} }
 							units={ units }
+							min={ MIN_FONT_SIZE }
+							max={ MAX_FONT_SIZE }
 						/>
 					</FlexItem>
 					<FlexItem isBlock>
@@ -152,6 +157,8 @@ function TagCloudEdit( { attributes, setAttributes, taxonomies } ) {
 								onFontSizeChange( 'largestFontSize', value );
 							} }
 							units={ units }
+							min={ MIN_FONT_SIZE }
+							max={ MAX_FONT_SIZE }
 						/>
 					</FlexItem>
 				</Flex>
```

</details>

amustaque97 replied:
sorry I misunderstood your point. Let me try this ☝️

amustaque97 committed:
update regex for unit attribute in `wp_tag_cloud` function

amustaque97 committed:
Update UnitControl  to have min/max check

ntsekouras commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -32,8 +41,26 @@ const MIN_TAGS = 1;
  */
 const MAX_TAGS = 100;
 
+const MIN_FONT_SIZE = 1;
```
Let's change this to something smaller to accommodate use cases with `em/rem`. 
```suggestion
const MIN_FONT_SIZE = 0.1;
```

ntsekouras reviewed (approved):
Thanks for all the iterations and your patience here @amustaque97! Great work! 💯 

I've left a minor change to the min value, but tested and this looks good! Attribute sanitization looks good too.

amustaque97 committed:
change `MIN_FONT_SIZE` to 0.1

Co-authored-by: Nik Tsekouras <ntsekouras@outlook.com>
